File: tools/lint/checks/agents.mjs
import { existsSync, readdirSync, readFileSync } from 'node:fs';
import { join, basename } from 'node:path';
import { parse } from 'yaml';
import {
  AGENT_KEYS, AGENT_REQUIRED_KEYS, MODEL_CLASSES, AGENT_TOOLS, AGENT_WORKSPACES,
} from '../../lib/constants.mjs';

export function agentFiles(cap) {
  const dir = join(cap.dir, 'agents');
  if (!existsSync(dir)) return [];
  return readdirSync(dir).filter((n) => n.endsWith('.agent.yaml')).map((n) => join(dir, n));
}

export function agentNames(cap) {
  return agentFiles(cap).map((f) => {
    try {
      return parse(readFileSync(f, 'utf8'))?.name;
    } catch {
      return null;
    }
  }).filter(Boolean);
}

// ARCHITECTURE §2.3 — the neutral agent spec carries only what all first-tier
// harnesses can express. Anything else belongs in adapters/<harness>/.
export function checkAgents({ caps, report }) {
  for (const cap of caps) {
    for (const path of agentFiles(cap)) {
      const file = `${cap.rel}/agents/${basename(path)}`;
      let data;
      try {
        data = parse(readFileSync(path, 'utf8'));
      } catch (e) {
        report('error', 'agent/parse', file, e.message);
        continue;
      }
      if (data == null || typeof data !== 'object') {
        report('error', 'agent/parse', file, 'agent spec is not a YAML mapping');
        continue;
      }
      for (const key of Object.keys(data)) {
        if (!AGENT_KEYS.includes(key)) {
          report('error', 'agent/unknown-key', file, `"${key}" is not in the §2.3 neutral vocabulary (no provider names, no harness tuning)`);
        }
      }
      for (const key of AGENT_REQUIRED_KEYS) {
        if (!data[key]) report('error', 'agent/required', file, `"${key}" is required`);
      }
      if (data.name && basename(path) !== `${data.name}.agent.yaml`) {
        report('error', 'agent/name-file', file, `name "${data.name}" must match filename ${data.name}.agent.yaml`);
      }
      if (data.model_class && !MODEL_CLASSES.includes(data.model_class)) {
        report('error', 'agent/model-class', file, `model_class must be one of {${MODEL_CLASSES.join(', ')}}`);
      }
      for (const tool of data.tools ?? []) {
        if (!AGENT_TOOLS.includes(tool)) {
          report('error', 'agent/tool', file, `tool "${tool}" is not in the neutral vocabulary {${AGENT_TOOLS.join(', ')}}`);
        }
      }
      if (data.workspace && !AGENT_WORKSPACES.includes(data.workspace)) {
        report('error', 'agent/workspace', file, `workspace must be one of {${AGENT_WORKSPACES.join(', ')}}`);
      }
      for (const ref of data.context_files ?? []) {
        if (!existsSync(join(cap.dir, ref))) {
          report('error', 'agent/context-file', file, `context_files entry "${ref}" does not resolve inside the capability`);
        }
      }
    }
  }
}

File: tools/aos_lint/src/aos_lint/checks/agents.py
import yaml

from ..constants import (
    AGENT_KEYS, AGENT_REQUIRED_KEYS, AGENT_TOOLS, AGENT_WORKSPACES, MODEL_CLASSES,
)


def agent_files(cap) -> list:
    directory = cap.dir / "agents"
    if not directory.exists():
        return []
    return [p for p in sorted(directory.iterdir(), key=lambda p: p.name)
            if p.name.endswith(".agent.yaml")]


def agent_names(cap) -> list[str]:
    """Every declared agent's `name:`, dropping the unreadable and the unnamed.

    Deliberately NOT `aos_cap.names.declared_agent_ids`: that one falls back to the filename
    stem, because a name COMPUTATION must never raise or come back empty. Here the absence is
    the finding — `agent/required` reports the missing `name:`, and a stem substituted behind
    its back would let `schedules/agent` and `skill/used-by-ref` resolve against a name the
    installed agent will not actually have."""
    out = []
    for path in agent_files(cap):
        try:
            data = yaml.safe_load(path.read_text(encoding="utf-8"))
        except (yaml.YAMLError, OSError):
            continue
        name = data.get("name") if isinstance(data, dict) else None
        if name:
            out.append(name)
    return out


# ARCHITECTURE §2.3 — the neutral agent spec carries only what all first-tier
# harnesses can express. Anything else belongs in adapters/<harness>/.
def check_agents(ctx) -> None:
    for cap in ctx.caps:
        for path in agent_files(cap):
            file = f"{cap.rel}/agents/{path.name}"
            try:
                data = yaml.safe_load(path.read_text(encoding="utf-8"))
            except yaml.YAMLError as e:
                ctx.report("error", "agent/parse", file, str(e))
                continue
            if not isinstance(data, dict):
                ctx.report("error", "agent/parse", file, "agent spec is not a YAML mapping")
                continue
            for key in data:
                if key not in AGENT_KEYS:
                    ctx.report("error", "agent/unknown-key", file,
                               f'"{key}" is not in the §2.3 neutral vocabulary (no provider '
                               f"names, no harness tuning)")
            for key in AGENT_REQUIRED_KEYS:
                if not data.get(key):
                    ctx.report("error", "agent/required", file, f'"{key}" is required')
            if data.get("name") and path.name != f"{data['name']}.agent.yaml":
                ctx.report("error", "agent/name-file", file,
                           f'name "{data["name"]}" must match filename '
                           f"{data['name']}.agent.yaml")
            if data.get("model_class") and data["model_class"] not in MODEL_CLASSES:
                ctx.report("error", "agent/model-class", file,
                           f"model_class must be one of {{{', '.join(MODEL_CLASSES)}}}")
            for tool in data.get("tools") or []:
                if tool not in AGENT_TOOLS:
                    ctx.report("error", "agent/tool", file,
                               f'tool "{tool}" is not in the neutral vocabulary '
                               f"{{{', '.join(AGENT_TOOLS)}}}")
            if data.get("workspace") and data["workspace"] not in AGENT_WORKSPACES:
                ctx.report("error", "agent/workspace", file,
                           f"workspace must be one of {{{', '.join(AGENT_WORKSPACES)}}}")
            for ref in data.get("context_files") or []:
                if not (cap.dir / str(ref)).exists():
                    ctx.report("error", "agent/context-file", file,
                               f'context_files entry "{ref}" does not resolve inside the '
                               f"capability")
